src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,208 @@
+                return ss::make_exception_future(exception);
+            } else {
+                return ss::make_ready_future<>();
+            }
+        });
+    }
+
+    Func _func;
+    std::exception_ptr exception{nullptr};
+    seastar::readable_eventfd _on_done;
+
+    std::chrono::milliseconds _timeout{};
+};
+
+class spsc_queue {
+public:
+    spsc_queue();
+
+    void close();
+
+    ss::future<> push(work_item* item);
+
+    work_item* pop();
+
+    bool empty();
+
+private:
+    mutex _push_mutex;
+    seastar::readable_eventfd _is_not_full;
+
+    std::atomic<bool> is_stopped{false};
+
+    boost::lockfree::spsc_queue<work_item*> _items;
+
+    std::mutex _std_mutex;
+    std::condition_variable _has_element_cv;
+    static constexpr std::chrono::milliseconds _timeout_cond_wait_std_thread_ms{
+      30};
+
+    ss::shard_id _ss_mutex_shard;
+};
+
+} // namespace internal
+
+// This class implement executor (std::thread) for runing v8 script. It
+// uses boost::spsc_queue for storing task.
+// For use we need create ss::shared<executor> and run submit like
+// executor_sharded.invoke_on(<core with executor>)
+class executor {
+public:
+    explicit executor(uint8_t cpu_id);
+
+    executor(const executor& other) = delete;
+    executor& operator=(const executor& other) = delete;
+    executor(executor&& other) = delete;
+    executor& operator=(executor&& other) = delete;
+
+    ~executor();
+
+    // Stop executor. Stop queue with task and close gate
+    ss::future<> stop();
+
+    bool is_stopping() const;
+
+    // clang-format off
+    CONCEPT(
+    template<typename T>
+    concept FuncFoExecutorConcept = requires(T func) {
+        { func() } -> void;
+        { func.cancel() } -> void;
+        { func.on_timeout() } -> void;
+    };)
+    // clang-format on
+
+    /// Submit new task in executor.
+    ///
+    /// \param start_func is used for set watchdog
+    /// \param func_for_executor is v8 script
+    template<typename WrapperFuncForExecutor>
+    CONCEPT(requires FuncFoExecutorConcept<WrapperFuncForExecutor>)
+    ss::future<> submit(
+      WrapperFuncForExecutor&& func_for_executor,
+      std::chrono::milliseconds timeout) {
+        gate_guard guard{_gate};
+
+        auto new_task
+          = std::make_unique<internal::task<WrapperFuncForExecutor>>(
+            std::forward<WrapperFuncForExecutor>(func_for_executor), timeout);
+
+        co_await _tasks.push(new_task.get());
+        co_await std::move(new_task->get_future());
+    }
+
+private:
+    // Set callback and rearm watchdog
+    void rearm_watchdog(
+      internal::work_item& task, std::chrono::milliseconds timeout);
+
+    // Cancel watchdog
+    void cancel_watchdog(internal::work_item& task);
+
+    // We need to pin thread to core without seastar reactor
+    void pin(unsigned cpu_id);
+
+    // Main loop for threads in executor
+    void loop();
+
+    std::atomic<bool> _is_stopped{false};
+
+    std::thread _threads;
```

Comment:
I would name it `_thread` since there is only one thread now.

Reply:
done